### PR TITLE
Count GPU usage

### LINKED ISSUE #51: [ColdFront/OpenStack] Auditing script for number of GPUs currently allocated on a project and quotas
The attribute for setting a quota on GPUs doesn't really do anything since OpenStack does not support any way to set a limit on that particular resource (pci-passthrough devices).

This script will just run periodically and
- count the number of active GPUs each project
- compare it to the quota
- if the project is above quota, send them and admins an email notification

### PR DESCRIPTION
**Pull request by knikolla**
Implements the command `count_gpu_usage --flavor <flavor_name>` which
counts the number of GPU instances and compares it with the allowance.

Closes #51

### CODE REVIEW

**Comment by knikolla** [comment]
According to https://github.com/nerc-project/nerc-docs/pull/66, there are now different types of GPU flavors. Given that this PR was written when there was only 1 GPU flavor and quota, it needs to be updated.

**Comment by joachimweyl** [comment]
> According to [nerc-project/nerc-docs#66](https://github.com/nerc-project/nerc-docs/pull/66), there are now different types of GPU flavors. Given that this PR was written when there was only 1 GPU flavor and quota, it needs to be updated.

good catch!

**Comment by knikolla** [comment]
@jtriley Reworked it to take `--flavor flavor1=2 --flavor flavor2=4`. I haven't had a chance to test it yet though, do you have any spare cycles to take that on?

**Comment by jtriley** [comment]
@knikolla I should be able to test by end of day tomorrow at the latest. I'll report back. Thanks!

File: src/coldfront_plugin_openstack/management/commands/count_gpu_usage.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            else:
+                logger.critical(
+                    f'Invalid value {flavor_arg}. Must be "<name>=<value>."'
+                )
+                sys.exit(1)
+
+        # Search the flavors, and construct a list of (flavor_obj, amount used by 1 instance)
+        flavors = []
+        for nova_flavor in client.flavors.list():
+            for name, index in enumerate(flavor_names):
```
Comment (jtriley) [inline src/coldfront_plugin_openstack/management/commands/count_gpu_usage.py:56]: The index comes first with the `enumerate` function so this should be:

```
for index, name in enumerate(flavor_names):
    ...
```

File: src/coldfront_plugin_openstack/management/commands/count_gpu_usage.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            help="Flavor of GPU quota instances in the form <flavor name>=<quota used by 1 instance>.",
+        )
+
+    def handle(self, *args, **options):
+        resource = Resource.objects.get(
+            resource_type=ResourceType.objects.get(name="OpenStack"),
+            name=options["resource"],
+        )
+
+        session = openstack.get_session_for_resource()
```
Comment (jtriley) [inline src/coldfront_plugin_openstack/management/commands/count_gpu_usage.py:37]: Need to pass in the resource object from the line before.

**Review by jtriley — CHANGES_REQUESTED** [review]

**Comment by jtriley** [comment]
Once I put in manual fixes for those two issues I hit (I put in some print statements to show where it's at in the loop):

```
root@coldfront-deployment-debug:~# coldfront count_gpu_usage --resource NERC --flavor gpu-a100.1=1
name: gpu-a100.1
nova_flavor.name: gpu-a100.2
name: gpu-a100.1
nova_flavor.name: mem-a.8   
name: gpu-a100.1
nova_flavor.name: cpu-a.1   
name: gpu-a100.1
nova_flavor.name: cpu-a.8   
name: gpu-a100.1
nova_flavor.name: mem-a.4   
name: gpu-a100.1
nova_flavor.name: mem-a.1   
name: gpu-a100.1
nova_flavor.name: gpu-a100.4
name: gpu-a100.1
nova_flavor.name: mem-a.16  
name: gpu-a100.1
nova_flavor.name: cpu-a.16  
name: gpu-a100.1
nova_flavor.name: cpu-a.4   
name: gpu-a100.1
nova_flavor.name: gpu-a100.1
name: gpu-a100.1
nova_flavor.name: mem-a.2   
name: gpu-a100.1
nova_flavor.name: cpu-a.2   
Traceback (most recent call last):
  File "/opt/venv/bin/coldfront", line 33, in <module>
    sys.exit(load_entry_point('coldfront==1.0.4', 'console_scripts', 'coldfront')())
  File "/opt/venv/lib/python3.9/site-packages/coldfront/__init__.py", line 11, in manage
    execute_from_command_line(sys.argv)
  File "/opt/venv/lib/python3.9/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/opt/venv/lib/python3.9/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/venv/lib/python3.9/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/venv/lib/python3.9/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/opt/venv/lib/python3.9/site-packages/coldfront_plugin_openstack/management/commands/count_gpu_usage.py", line 67, in handle
    allocations = Allocation.objects.filter(
  File "/opt/venv/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/opt/venv/lib/python3.9/site-packages/django/db/models/query.py", line 941, in filter
    return self._filter_or_exclude(False, args, kwargs)
  File "/opt/venv/lib/python3.9/site-packages/django/db/models/query.py", line 961, in _filter_or_exclude
    clone._filter_or_exclude_inplace(negate, args, kwargs)
  File "/opt/venv/lib/python3.9/site-packages/django/db/models/query.py", line 968, in _filter_or_exclude_inplace
    self._query.add_q(Q(*args, **kwargs))
  File "/opt/venv/lib/python3.9/site-packages/django/db/models/sql/query.py", line 1393, in add_q
    clause, _ = self._add_q(q_object, self.used_aliases)
  File "/opt/venv/lib/python3.9/site-packages/django/db/models/sql/query.py", line 1412, in _add_q
    child_clause, needed_inner = self.build_filter(
  File "/opt/venv/lib/python3.9/site-packages/django/db/models/sql/query.py", line 1347, in build_filter
    condition = self.build_lookup(lookups, col, value)
  File "/opt/venv/lib/python3.9/site-packages/django/db/models/sql/query.py", line 1193, in build_lookup
    lookup = lookup_class(lhs, rhs)
  File "/opt/venv/lib/python3.9/site-packages/django/db/models/lookups.py", line 25, in __init__
    self.rhs = self.get_prep_lookup()
  File "/opt/venv/lib/python3.9/site-packages/django/db/models/fields/related_lookups.py", line 50, in get_prep_lookup
    self.rhs = [get_normalized_value(val, self.lhs)[0] for val in self.rhs]
TypeError: 'Resource' object is not iterable
```

File: src/coldfront_plugin_openstack/management/commands/count_gpu_usage.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                if nova_flavor.name == name:
+                    flavors.append((nova_flavor, flavor_values[index]))
+        if len(flavors) != len(flavor_names):
+            logger.critical(f"Not all flavor names found!")
+            sys.exit(1)
+
+        # Find all active OpenStack projects and create a
+        # dictionary with the project id
+        allocations = Allocation.objects.filter(
+            resources__in=resource,
```
Comment (jtriley) [inline src/coldfront_plugin_openstack/management/commands/count_gpu_usage.py:66]: This needs to be a list, e.g:

```
        allocations = Allocation.objects.filter(
            resources__in=[resource],
            ...
```

**Review by jtriley — CHANGES_REQUESTED** [review]

File: src/coldfront_plugin_openstack/management/commands/count_gpu_usage.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                    f" {project_id} containing {active_gpu_count} GPU instances."
+                )
+                logger.error(msg)
+                continue
+
+            allowed = allocation.get_attribute(attributes.QUOTA_GPU)
+            if active_gpu_count > allowed:
+                msg = (
+                    f'{allocation.pk} of project "{allocation.project.title}"'
+                    f" is using {active_gpu_count} GPU instances. (Allowed {allowed}."
```
Comment (jtriley) [inline src/coldfront_plugin_openstack/management/commands/count_gpu_usage.py:104]: Missing end-quote in the message here, e.g.:

```
                   f" is using {active_gpu_count} GPU instances. (Allowed {allowed})."
```

**Review by jtriley — CHANGES_REQUESTED** [review]

**Comment by jtriley** [comment]
Will try some other flavors soon and verify the results

**Comment by knikolla** [comment]
Thank you for massaging this to work, and sorry for not having the time to properly test it. I'll revise with the fixes tomorrow, unless you'd like to do that before then.

**Comment by jtriley** [comment]
Yeah I think just better phrasing like what you suggested would be good. I can implement all of these changes and have you review. Thanks!

**Comment by jtriley** [comment]
Ah on second thought I don't have access to your branch so I'll let you do it :)

**Comment by jtriley** [comment]
🤷 turns out github let me make the changes in your branch so all of this has been implemented. Going to update the option parsing to assume GPU count is 1 if no count is specified and log that it's doing that via debug message.

**Review by jtriley — APPROVED** [review]

**Comment by jtriley** [comment]
@knikolla I implemented all of the fixes discussed here in your branch. Please take a look and feel free to merge if it looks good to you.

**Comment by knikolla** [comment]
Thanks!

**Pull request merged**